src/plugindata.h

```diff
@@ -263,8 +265,66 @@ void plugin_configure_single(GtkWidget *parent);
 void plugin_help(void);
 void plugin_cleanup(void);
 
+/** Called by Geany when a plugin library is loaded.
+ *
+ * This is the original entry point. Implement and export this function to be loadable at all.
+ * Then fill in GeanyPlugin::info and GeanyPlugin::funcs of the passed @p plugin. Finally
+ * GEANY_PLUGIN_REGISTER() and specify a minimum supported API version.
+ *
+ * For all glory details please read @ref howto.
+ *
+ * @param plugin The unique plugin handle to your plugin. You must set some fields here.
+ * @param module The GModule associated with your plugin library.
+ * @param geany_api_version The API version exposed by the Geany instance that loads your plugin.
+ *           This probably not the same as the one your plugin was compiled against.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ */
+void geany_load_module(GeanyPlugin *plugin, GModule *module, gint geany_api_version);
+
 #endif
 
+/** Callback functions that need to be implemented for every plugin.
+ *
+ * These callbacks should be registered by the plugin within Geany's call to
+ * geany_load_module() by calling geany_plugin_register() with an instance of this type.
+ *
+ * Geany will then call the callbacks at appropriate times. Each gets passed the
+ * plugin-defined data pointer as well as the corresponding GeanyPlugin instance
+ * pointer.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ **/
+struct GeanyPluginFuncs
+{
+	/** Array of plugin-provided signal handlers @see PluginCallback */
+	PluginCallback *callbacks;
+	/** Called to initialize the plugin, when the user activates it (must not be @c NULL) */
+	gboolean    (*init)      (GeanyPlugin *plugin, gpointer pdata);
+	/** plugins configure dialog, optional (can be @c NULL) */
+	GtkWidget*  (*configure) (GeanyPlugin *plugin, GtkDialog *dialog, gpointer pdata);
+	/** Called when the plugin should show some help, optional (can be @c NULL) */
+	void        (*help)      (GeanyPlugin *plugin, gpointer pdata);
+	/** Called when the plugin is disabled or when Geany exits (must not be @c NULL) */
```

Comment:
Am 16.08.2015 um 17:05 schrieb Colomban Wendling:
>
> Well, if |set_data()| can be called from |init()| it basically have to 
> get cleaned up after |cleanup()| otherwise it either doesn't work 
> (with the current one-time-only |set_data()|), or isn't any good (as 
> the data would be unexpectedly kept after the plugin got cleaned up).
> And if |set_data()| really can only be called from 
> |geany_load_module()| it should be enforced in some way, and made 
> clear from the docs.
>

Actually the _set_data() is even a separate function only so that you 
can call it from init(). [And with proxy plugins geany_load_module() 
isn't necessarily called for each plugin.]

>     Actually, I think we'd have to call the free_func even then,
>     because otherwise it would mean geany_plugin_set_data() would work
>     on the first run of init() but not subsequently (remember I
>     intentionally limited to setting data only once).
>
> But then you'd have to re-call |geany_load_module()| anyway, because 
> if |set_data()| came from it you need to let the plugin re-set it.
>
>

Right. Humm...